torch/distributions/constraints.py

```diff
@@ -297,6 +301,59 @@ def check(self, value):
         return (value == value).all()  # False for NANs.
 
 
+class _Cat(Constraint):
+    """
+    Constraint functor that applies a sequence of constraints
+    `cseq` at the submatrices at dimension `dim`,
+    each of size `lengths[dim]`, in a way compatible with :func:`torch.cat`.
+    """
+    def __init__(self, cseq, dim=0, lengths=None):
+        assert all(isinstance(c, Constraint) for c in cseq)
+        self.cseq = list(cseq)
+        if lengths is None:
+            lengths = [1] * len(self.cseq)
+        self.lengths = list(lengths)
+        assert len(self.lengths) == len(self.cseq)
+        self.dim = dim
+
+    def check(self, value):
+        assert -value.dim() <= self.dim < value.dim()
+        checks = []
+        start = 0
+        for constr, length in zip(self.cseq, self.lengths):
+            v = value.narrow(self.dim, start, length)
+            checks.append(constr.check(v))
+            start = start + length  # avoid += for jit compat
+        return torch.cat(checks, self.dim)
+
+
+class _Stack(Constraint):
+    """
+    Constraint functor that applies a sequence of constraints
+    `cseq` at the submatrices at dimension `dim`,
+    each of size `lengths[dim]` in a way compatible with :func:`torch.stack`.
+    """
+    def __init__(self, cseq, dim=0, lengths=None):
```

Comment:
Hmm I think `lengths` makes sense for `Cat` but not for `Stack`, right? IIUC `torch.stack` creates a new dimension, so the size of each thing in the new dimension will be 1.

Reply:
I think I agree! I was just thinking about how weird it was when I was doing the slicing.